Refresh.GameServer/Database/GameDatabaseContext.Playlists.cs

```diff
@@ -101,56 +170,165 @@ public void AddLevelToPlaylist(GameLevel level, GamePlaylist parent)
-
-            if (relation == null)
-                return;
-            
             this.LevelPlaylistRelations.Remove(relation);
         });
     }
 
+    private void DecreasePlaylistLevelIndicesAfterIndex(GamePlaylist playlist, int startIndex)
+    {
+        IEnumerable<LevelPlaylistRelation> relations = this.LevelPlaylistRelations
+            .Where(r => r.Playlist == playlist && r.Index >= startIndex)
+            .AsEnumerable();
+
+        this.Write(() => {
+            foreach(LevelPlaylistRelation relation in relations)
+            {
+                relation.Index--;
+            }
+        });
+    }
+
+    public void SetPlaylistLevelIndex(GamePlaylist playlist, GameLevel level, int newIndex)
+    {
+        LevelPlaylistRelation relation = this.LevelPlaylistRelations
+            .First(r => r.Playlist == playlist && r.Level == level);
+
+        this.Write(() => {
+            relation.Index = newIndex;
+        });
+    }
+
     public IEnumerable<GamePlaylist> GetPlaylistsContainingPlaylist(GamePlaylist playlist)
         // TODO: with postgres this can be IQueryable
         => this.SubPlaylistRelations.Where(p => p.SubPlaylist == playlist).AsEnumerable()
+            .Reverse()
             .Select(r => this.GamePlaylists.First(p => p.PlaylistId == r.Playlist.PlaylistId))
             .Where(p => !p.IsRoot);
 
     public IEnumerable<GamePlaylist> GetPlaylistsByAuthorContainingPlaylist(GameUser user, GamePlaylist playlist)
         // TODO: with postgres this can be IQueryable
         => this.SubPlaylistRelations.Where(p => p.SubPlaylist == playlist).AsEnumerable()
+            .Reverse()
             .Select(r => this.GamePlaylists.First(p => p.PlaylistId == r.Playlist.PlaylistId))
             .Where(p => p.Publisher.UserId == user.UserId)
             .Where(p => !p.IsRoot);
 
-    public IEnumerable<GameLevel> GetLevelsInPlaylist(GamePlaylist playlist, TokenGame game) =>
+    public IEnumerable<GameLevel> GetLevelsInPlaylist(GamePlaylist playlist, TokenGame game)
         // TODO: When we have postgres, remove the `AsEnumerable` call for performance. 
-        this.LevelPlaylistRelations.Where(l => l.Playlist == playlist).AsEnumerable()
+        => this.LevelPlaylistRelations
+            .Where(l => l.Playlist == playlist)
+            .OrderBy(r => r.Index)
+            .AsEnumerable()
             .Select(l => l.Level)
             .FilterByGameVersion(game);
 
+    public int GetTotalLevelsInPlaylistCount(GamePlaylist playlist, TokenGame game) => 
+        this.LevelPlaylistRelations.Where(l => l.Playlist == playlist).AsEnumerable()
+            .Select(l => l.Level)
+            .FilterByGameVersion(game)
+            .Count();
+
+    public int GetTotalLevelsInPlaylistCount(GamePlaylist playlist) => 
+        this.LevelPlaylistRelations.Where(l => l.Playlist == playlist).AsEnumerable()
+            .Select(l => l.Level)
+            .Count();
```

Comment:
Since the level itself is unused, this might be able to be shortened to this:

```suggestion
    public int GetTotalLevelsInPlaylistCount(GamePlaylist playlist) => 
        this.LevelPlaylistRelations.Count(l => l.Playlist == playlist)
```